oadp-operator/controllers/velero.go

```diff
@@ -1,14 +1,486 @@
+	VELERO_AZURE_SECRET_NAME = "cloud-credentials-azure"
+	VELERO_GCP_SECRET_NAME   = "cloud-credentials-gcp"
+)
+
+// Environment Vars
+const (
+	LD_LIBRARY_PATH               = "LD_LIBRARY_PATH"
+	VELERO_NAMESPACE              = "VELERO_NAMESPACE"
+	VELERO_SCRATCH_DIR            = "VELERO_SCRATCH_DIR"
+	AWS_SHARED_CREDENTIALS_FILE   = "AWS_SHARED_CREDENTIALS_FILE"
+	AZURE_SHARED_CREDENTIALS_FILE = "AZURE_SHARED_CREDENTIALS_FILE"
+	GCP_SHARED_CREDENTIALS_FILE   = "GCP_SHARED_CREDENTIALS_FILE"
+	//TODO: Add Proxy env vars
+	HTTP_PROXY  = "HTTP_PROXY"
+	HTTPS_PROXY = "HTTPS_PROXY"
+	NO_PROXY    = "NO_PROXY"
+)
+
+//TODO: Add Image customization options
+// Images
+const (
+	VELERO_IMAGE           = "quay.io/konveyor/velero:latest"
+	OPENSHIFT_PLUGIN_IMAGE = "quay.io/konveyor/openshift-velero-plugins:latest"
+	AWS_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-aws:latest"
+	AZURE_PLUGIN_IMAGE     = "quay.io/konveyor/velero-plugin-for-microsoft-azure:latest"
+	GCP_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-gcp:latest"
+	CSI_PLUGIN_IMAGE       = "quay.io/konveyor/velero-plugin-for-csi:latest"
+)
+
+// Plugin names
+const (
+	VELERO_PLUGIN_FOR_AWS       = "velero-plugin-for-aws"
+	VELERO_PLUGIN_FOR_AZURE     = "velero-plugin-for-microsoft-azure"
+	VELERO_PLUGIN_FOR_GCP       = "velero-plugin-for-gcp"
+	VELERO_PLUGIN_FOR_CSI       = "velero-plugin-for-csi"
+	VELERO_PLUGIN_FOR_OPENSHIFT = "openshift-velero-plugin"
 )
 
 func (r *VeleroReconciler) ReconcileVeleroDeployment(log logr.Logger) (bool, error) {
 	velero := oadpv1alpha1.Velero{}
 	if err := r.Get(r.Context, r.NamespacedName, &velero); err != nil {
 		return false, err
 	}
+
+	veleroVolumeMounts := r.getVeleroVolumeMounts(&velero)
+	veleroVolumes := r.getVeleroVolumes(&velero)
+	veleroEnv := r.getVeleroEnv(&velero)
+	veleroInitContainers := r.getVeleroInitContainers(&velero)
+	veleroResourceReqs := r.getVeleroResourceReqs(&velero)
+	veleroTolerations := velero.Spec.VeleroTolerations
+	veleroLabels := r.getAppLabels(&velero)
+	veleroDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      VELERO,
+			Namespace: VELERO_NS,
+		},
+	}
+
+	op, err := controllerutil.CreateOrUpdate(r.Context, r.Client, veleroDeployment, func() error {
+
+		// Setting Deployment selector if a new object is created as it is immutable
+		if veleroDeployment.ObjectMeta.CreationTimestamp.IsZero() {
+			veleroDeployment.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"component": VELERO,
+				},
+			}
+		}
+
+		// Setting controller owner reference on the velero deployment
+		err := controllerutil.SetControllerReference(&velero, veleroDeployment, r.Scheme)
+		if err != nil {
+			return err
+		}
+
+		// update the Deployment template
+		veleroDeployment = r.buildVeleroDeployment(veleroDeployment, veleroLabels, veleroVolumeMounts, veleroVolumes, veleroEnv, veleroInitContainers, veleroResourceReqs, veleroTolerations)
```

Comment:
an opinionated suggestion: this function call is not providing a lot of values, you are sending the arg list into the function and it is just setting it to the right place in deployment. 

Why don't we change this to `r.buildVeleroDeployment(veleroDeployment)`? then the function will call all the getters and build the right spec there. Then we can add a unit test for `buildVeleroDeployment`. 

Not asking for the units in this PR but let's start with refactoring this function so it more makes

Reply:
Agree with this. Generally my rule of thumb is anything over 3 args means we can refactor the function call in some manner. I think Alay's suggestion makes sense